### PR TITLE
Retrofit

### PR DESCRIPTION
**Pull request by tolia-ogurt**

### CODE REVIEW

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 1]
```diff
@@ -1,19 +1,62 @@
...
+import com.example.trempel.network.ServiceProvider
+import com.example.trempel.network.model.ProductModel
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
 
 internal class PdpFragment : Fragment() {
 
+    private var _binding: PdpFragmentBinding? = null
+    private val binding get() = _binding!!
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: can we remove this approach with `!!`?

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 2]
```diff
@@ -1,19 +1,62 @@
...
         savedInstanceState: Bundle?
-    ): View? {
-        return inflater.inflate(R.layout.pdp_fragment, container, false)
+    ): View {
+        _binding = PdpFragmentBinding.inflate(inflater, container, false)
+        getInfoProduct()
+        return binding.root
     }
 
+    private fun getInfoProduct() {
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: I'd recommend `getProductInfo`

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 3]
```diff
@@ -1,19 +1,62 @@
...
+        ServiceProvider.getRetrofitService()?.getProduct()
+            ?.enqueue(object : Callback<ProductModel> {
+                @SuppressLint("SetTextI18n")
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: Let's not suppress lint warnings but try to resolve it

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 4]
```diff
@@ -1,19 +1,62 @@
...
+                override fun onResponse(
+                    call: Call<ProductModel>,
+                    response: Response<ProductModel>
+                ) {
+                    binding.tvProductTitle.text = response.body()?.title
+                    Glide.with(requireContext()).load(response.body()?.image)
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: I don't like the idea of using `requireContext()`, let's change it to null safe call? 
_Probably we can use view context for it_

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 5]
```diff
@@ -1,19 +1,62 @@
...
+                        .into(binding.ivProduct)
+                    binding.tvProductPrice.text = "$${response.body()?.price.toString()}"
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: Do we need `$$`?

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 6]
```diff
@@ -1,19 +1,62 @@
...
 
+    private fun getInfoProduct() {
+        ServiceProvider.getRetrofitService()?.getProduct()
+            ?.enqueue(object : Callback<ProductModel> {
+                @SuppressLint("SetTextI18n")
+                override fun onResponse(
+                    call: Call<ProductModel>,
+                    response: Response<ProductModel>
+                ) {
+                    binding.tvProductTitle.text = response.body()?.title
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: I'd suggest to use `with` scope function

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 7]
```diff
@@ -1,19 +1,62 @@
...
+                    Glide.with(requireContext()).load(response.body()?.image)
+                        .into(binding.ivProduct)
+                    binding.tvProductPrice.text = "$${response.body()?.price.toString()}"
+                    binding.tvProductDescription.text = response.body()?.description
+                    binding.rbProductRating.rating = response.body()?.rating?.rate?.toFloat() ?: 0f
+                    binding.tvProductScore.text = response.body()?.rating?.rate.toString()
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: Are you sure that we should call `toString()` ignoring null safety?
_Try to call `.toString()` on `null`, what result do you have?_

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 8]
```diff
@@ -1,19 +1,62 @@
...
+                    binding.tvReviews.text = "(${response.body()?.rating?.count.toString()})"
+                }
+
+                override fun onFailure(call: Call<ProductModel>, t: Throwable) {
+                    Log.e(this.javaClass.name, t.toString())
+                    Toast.makeText(
+                        requireContext(),
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: let's change it to null safe call

https://medium.com/@shafran/fragment-getcontext-vs-requirecontext-ffc9157d6bbe

File: app/src/main/java/com/example/trempel/network/ServiceProvider.kt [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package com.example.trempel.network
+
+import com.example.trempel.network.service.RetrofitService
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+
+object ServiceProvider {
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/network/ServiceProvider.kt]: can we change to `internal`?

File: app/src/main/java/com/example/trempel/network/model/ProductModel.kt [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package com.example.trempel.network.model
+
+data class ProductModel(
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/network/model/ProductModel.kt]: `internal`?

File: app/src/main/java/com/example/trempel/network/model/Rating.kt [hunk 11]
```diff
@@ -0,0 +1,6 @@
+package com.example.trempel.network.model
+
+data class Rating(
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/network/model/Rating.kt]: `internal`?

File: app/src/main/java/com/example/trempel/network/model/Rating.kt [hunk 12]
```diff
@@ -0,0 +1,6 @@
+package com.example.trempel.network.model
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/network/model/Rating.kt]: We can have in the same file as `ProductModel` if you like

File: app/src/main/java/com/example/trempel/network/service/RetrofitService.kt [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package com.example.trempel.network.service
+
+import com.example.trempel.network.model.ProductModel
+import retrofit2.Call
+import retrofit2.http.GET
+
+interface RetrofitService {
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/network/service/RetrofitService.kt]: `internal`?

File: app/src/main/java/com/example/trempel/network/service/RetrofitService.kt [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
+    @GET("/products/2")
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/network/service/RetrofitService.kt]: Let's have `2` as parameter?

File: app/src/main/res/layout/pdp_fragment.xml [hunk 15]
```diff
@@ -17,18 +17,18 @@
...
-        android:src="@drawable/image_product" />
+        tools:src="@drawable/image_product" />
 
     <View
         android:id="@+id/view"
         android:layout_width="0dp"
         android:layout_height="0dp"
-        app:layout_constraintTop_toBottomOf="@id/iv_product"
+        android:layout_marginTop="@dimen/normal_150"
+        android:background="@drawable/ic_rectangle_1"
```
Comment (DaryaErofeeva) [inline app/src/main/res/layout/pdp_fragment.xml]: can we rename `rectangle_1` to something more concrete?
E.g. `pdp_rounded_background` or other option that you like

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 2 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: done

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 3 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: I can not clear this warnings

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 6 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: done

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 7 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: done

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 8 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: done

File: app/src/main/java/com/example/trempel/network/ServiceProvider.kt (hunk 9 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/network/ServiceProvider.kt]: done

File: app/src/main/java/com/example/trempel/network/model/ProductModel.kt (hunk 10 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/network/model/ProductModel.kt:3]: done

File: app/src/main/java/com/example/trempel/network/model/Rating.kt (hunk 11 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/network/model/Rating.kt:3]: done

File: app/src/main/java/com/example/trempel/network/service/RetrofitService.kt (hunk 14 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/network/service/RetrofitService.kt]: @Path?

File: app/src/main/res/layout/pdp_fragment.xml (hunk 15 above)
Comment (tolia-ogurt) [reply app/src/main/res/layout/pdp_fragment.xml]: done

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 4 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: context!! maybe?

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 1 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: done

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 3 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: done

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 4 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: done

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 5 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: change

File: app/src/main/java/com/example/trempel/network/model/ProductModel.kt (hunk 10 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/network/model/ProductModel.kt]: done

File: app/src/main/java/com/example/trempel/network/model/Rating.kt (hunk 11 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/network/model/Rating.kt]: done

File: app/src/main/java/com/example/trempel/network/model/Rating.kt (hunk 12 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/network/model/Rating.kt]: done

File: app/src/main/res/layout/pdp_fragment.xml (hunk 15 above)
Comment (tolia-ogurt) [reply app/src/main/res/layout/pdp_fragment.xml]: done

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 16]
```diff
@@ -1,19 +1,81 @@
...
+            ?.enqueue(object : Callback<ProductModel> {
+                override fun onResponse(
+                    call: Call<ProductModel>,
+                    response: Response<ProductModel>
+                ) {
+
+                    binding?.run {
+                        tvProductTitle.text = response.body()?.title
+                        Glide.with(ivProduct).load(response.body()?.image)
+                            .into(ivProduct)
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: can fit one line

File: app/src/main/java/com/example/trempel/PdpFragment.kt [hunk 17]
```diff
@@ -1,19 +1,81 @@
...
+                                response.body()?.rating?.rate?.toString().orEmpty().length,
+                                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
+                            )
+                        }
+                    }
+                }
+
+                override fun onFailure(call: Call<ProductModel>, t: Throwable) {
+                    Log.e(this.javaClass.name, t.toString())
+                    Toast.makeText(context, "Something went wrong, try later.", Toast.LENGTH_LONG)
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/PdpFragment.kt]: magic string

File: app/src/main/java/com/example/trempel/network/ServiceProvider.kt [hunk 18]
```diff
@@ -0,0 +1,19 @@
+package com.example.trempel.network
+
+import com.example.trempel.network.service.RetrofitService
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+
+internal object ServiceProvider {
+
+    private const val BASE_URL = "https://fakestoreapi.com/"
+    private var retrofit: Retrofit? = null
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/network/ServiceProvider.kt]: do we really need `retrofit` defined here?

File: app/src/main/java/com/example/trempel/network/model/ProductModel.kt [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+    val id: Int,
+    val image: String,
+    val price: Double,
+    val rating: Rating,
+    val title: String
+)
+
+internal data class Rating(
+    val count: Int,
+    val rate: Double
```
Comment (DaryaErofeeva) [inline app/src/main/java/com/example/trempel/network/model/ProductModel.kt]: can we make it float?

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 16 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: done

File: app/src/main/java/com/example/trempel/PdpFragment.kt (hunk 17 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/PdpFragment.kt]: done

File: app/src/main/java/com/example/trempel/network/ServiceProvider.kt (hunk 18 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/network/ServiceProvider.kt]: done

File: app/src/main/java/com/example/trempel/network/model/ProductModel.kt (hunk 19 above)
Comment (tolia-ogurt) [reply app/src/main/java/com/example/trempel/network/model/ProductModel.kt]: done

File: app/build.gradle [hunk 20]
```diff
@@ -26,11 +27,20 @@ android {
...
+    def retrofit_version = "2.9.0"
+    def glide_version = "4.12.0"
 
     //Navigation
     implementation("androidx.navigation:navigation-fragment-ktx:$nav_version")
     implementation("androidx.navigation:navigation-ui-ktx:$nav_version")
 
+    //Retrofit
+    implementation "com.squareup.retrofit2:retrofit:$retrofit_version"
+    implementation 'com.squareup.retrofit2:converter-gson:2.5.0'
```
Comment (DaryaErofeeva) [inline app/build.gradle:39]: missed version extraction

**Review by DaryaErofeeva — APPROVED** [review]

**Pull request merged**